xla/pjrt/gpu/se_gpu_pjrt_client.cc

```diff
@@ -720,47 +722,60 @@ GetStreamExecutorGpuDeviceAllocator(
     se::Platform* platform, const GpuAllocatorConfig& allocator_config,
     const std::map<int, std::unique_ptr<LocalDeviceState>>&
         addressable_devices) {
-  std::unique_ptr<se::DeviceMemoryAllocator> allocator;
+  std::vector<se::MultiDeviceAdapter::AllocatorInfo> allocators;
   switch (allocator_config.kind) {
     case GpuAllocatorConfig::Kind::kCudaAsync: {
-      auto allocator_or = CreateCudaAsyncAllocator(
+      auto allocators_or = CreateCudaAsyncAllocator(
           platform, addressable_devices, allocator_config.memory_fraction,
           allocator_config.preallocate);
-      if (allocator_or.ok()) {
+      if (allocators_or.ok()) {
         LOG(INFO) << "Using CUDA async allocator.";
-        allocator = std::move(allocator_or.value());
+        allocators = std::move(allocators_or.value());
         break;
       }
       LOG(ERROR) << "Failed to initialize CUDA async allocator: "
-                 << allocator_or.status() << "; falling back to BFC.";
+                 << allocators_or.status() << "; falling back to BFC.";
       [[fallthrough]];
     }
 
     case GpuAllocatorConfig::Kind::kDefault:
     case GpuAllocatorConfig::Kind::kBFC: {
       LOG(INFO) << "Using BFC allocator.";
-      std::vector<se::MultiDeviceAdapter::AllocatorWithStream>
-          allocators_and_streams;
       for (const auto& ordinal_and_device : addressable_devices) {
         TF_ASSIGN_OR_RETURN(
             auto bfc_allocator,
             CreateBFCAllocator(ordinal_and_device.second->executor(),
                                allocator_config.memory_fraction,
                                allocator_config.preallocate));
-        allocators_and_streams.emplace_back(
-            std::move(bfc_allocator),
-            ordinal_and_device.second->compute_stream());
+        allocators.emplace_back(std::move(bfc_allocator),
+                                ordinal_and_device.second->compute_stream(),
+                                /*memory_space=*/0);
       }
-      allocator = std::make_unique<se::MultiDeviceAdapter>(
-          platform, std::move(allocators_and_streams));
       break;
     }
 
     case GpuAllocatorConfig::Kind::kPlatform:
       LOG(INFO) << "Using platform allocator.";
```

Comment:
JAX tests are currently failing in case of Platform allocator usage, since in this case the function returns MultiDeviceAdapter with empty list of allocators. Then when clients (such as StreamExecutorGpuDevice::GetAllocatorStats) calls GetAllocator/GetStream/Allocate we are getting exception at line memory_space_to_per_device_allocators_[0][device_ordinal] (the allocator for device was not provided so vector is empty).

The problem is that the contract of the MultiDeviceAdapter methods changed here in case of platform allocator usage.
Previously they were calling underlying methods:
https://github.com/openxla/xla/blob/main/xla/stream_executor/integrations/tf_allocator_adapter.h#L140

And then each underlying method could decide what to do in this case.

I'm trying to find a good way to fix this issue.

Reply:
It looks previously `GetStreamExecutorGpuDeviceAllocator` returns `nullptr` when kPlatform is used which then causes the allocator to be set here: https://github.com/openxla/xla/blob/main/xla/pjrt/pjrt_stream_executor_client.cc#L241 which ultimately uses `StreamExecutorMemoryAllocator`.

If we return `nullptr` for kPlatform, then it should fix the error but we won't be able to use collective memory. I think this would be okay for now?

Reply:
Just pushed a commit with ^, let me know your thoughts and if it fixes the tests.

Reply:
This solution solves only part of problem. Another issue is that there are many places in XLA where allocator is down-casted to the MultiDeviceAdapter (https://github.com/openxla/xla/blob/fa8ce5f237a289cd415fe39dac0a40d9c3d08321/xla/pjrt/gpu/se_gpu_pjrt_client.cc#L908).
In case when platform allocator is used eventually pjrt_client uses backend allocator (https://github.com/openxla/xla/blob/main/xla/pjrt/pjrt_stream_executor_client.cc#L241) which is actually StreamExectorMemoryAllocator (https://github.com/openxla/xla/blob/main/xla/service/backend.h#L199).

So this problem was even before the current change, but it was working because after casting the underlying StreamExectorMemoryAllocator to MultiDeviceAdapter the client could still use an underlying allocator.

I'm trying figure it out how to fix this situation.

Reply:
Actually seems like the only place were this cast should be fixed is StreamExecutorGpuDevice::GetAllocatorStats